File: project/src/main/resources/static/js/action/taskAction.js
import { socketSend } from "../app.js";
import { getUsername, checkMemberId } from "../common/common.js";

function createTask(projectId) {
    let name = $('#task-name').val();
    let content = $('#task-content').val();
    let teamId = $('#member-list').attr('teamid')
    let owner = $('#task-owner').val()
    let ownerId = checkMemberId(owner, teamId)
    let requester = $('#task-requester').val()
    let requesterId = checkMemberId(requester, teamId)
    let start = $('#task-start').val()
    let end = $('#task-end').val()
    let complexity = $('#task-complex').val()

    let message = {"cmd": "createTask",
        "message": {"projectId": projectId, "name": name, "content": content, "ownerId": ownerId,
            "requesterId": requesterId, "start": start,
            "end": end, "complexity": complexity, "username": getUsername()}}
    socketSend(message)
}

function updateTask(taskId, teamId) {
    let table = $('#update-table')
    let owner = table.find('#owner').val()
    owner = checkMemberId(owner, teamId)

    let message = {"cmd": "updateTask",
        "message": {"taskId": taskId, "name": table.find('#name').val(),
            "content": table.find('#content').val(), "ownerId": owner,
            "start": table.find('#start').val(), "end": table.find('#end').val(),
            "complexity": table.find('#complexity').val(), "username": getUsername()}}
    socketSend(message)
}

function deleteTask(taskId) {
    let message = {"cmd": "deleteTask",
        "message": {"taskId": taskId, "username": getUsername()}}
    socketSend(message)
}

function updateTaskStatus(taskId, content) {
    let status = Number(content) + 1
    let message = {"cmd": "updateTaskStatus",
        "message": {"taskId": taskId, "username": getUsername(), "status": status}}
    socketSend(message)
}

export { createTask, updateTask, deleteTask, updateTaskStatus }

File: project/src/main/java/com/cs633/project/action/TaskAction.java
package com.cs633.project.action;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONObject;
import com.cs633.project.IDataBus;
import com.cs633.project.constant.CommonConstant;
import com.cs633.project.message.Response;
import com.cs633.project.net.annotation.Action;
import com.cs633.project.utils.GetCurrentUser;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;

/**
 * Name: Donghang He
 * Date: 2021/10/27 4:45 下午
 * Course: CS-
 * Assignment
 * Description:
 */
@Action
@Service
public class TaskAction {

    @Autowired
    private IDataBus dataBus1;

    private static IDataBus dataBus;

    @PostConstruct
    public void init() {
        dataBus = dataBus1;
    }

    public JSONObject createTask(JSONObject object) {
        Long projectId = object.getLong("projectId");
        if (projectId == null) {
            return Response.sendErrorMessage(CommonConstant.ERROR_PROJECT_NOT_EXIST, "createTask");
        }

        return dataBus.taskService().createTask(GetCurrentUser.getUser(object), projectId, object.getString("name"),
                object.getString("content"), object.getLong("ownerId"), object.getLong("requesterId"),
                object.getDate("start"), object.getDate("end"), object.getInteger("complexity"));
    }

    public JSONObject updateTask(JSONObject object) {
        Long taskId = object.getLong("taskId");
        if (taskId == null) {
            return Response.sendErrorMessage(CommonConstant.ERROR_TASK_NOT_EXIST, "updateTask");
        }

        return dataBus.taskService().updateTask(GetCurrentUser.getUser(object), taskId, object.getString("name"),
                object.getString("content"), object.getLong("ownerId"), object.getLong("requesterId"),
                object.getDate("start"), object.getDate("end"), object.getInteger("complexity"));
    }

    public JSONObject deleteTask(JSONObject object) {
        Long taskId = object.getLong("taskId");
        if (taskId == null) {
            return Response.sendErrorMessage(CommonConstant.ERROR_TASK_NOT_EXIST, "deleteTask");
        }

        return dataBus.taskService().deleteTask(GetCurrentUser.getUser(object), taskId);
    }

    public JSONObject updateTaskStatus(JSONObject object) {
        Long taskId = object.getLong("taskId");
        if (taskId == null) {
            return Response.sendErrorMessage(CommonConstant.ERROR_TASK_NOT_EXIST, "updateTask");
        }

        return dataBus.taskService().updateTaskStatus(GetCurrentUser.getUser(object), taskId,
                object.getInteger("status"));
    }
}
